### PR TITLE
feat(Breadcrumb): Convert Breadcrumb jsx file to tsx

### PR DESCRIPTION
**Pull request by jessiehuff**
Fixes issue #1717 by updating breadcrumb to typescript.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -62,7 +62,7 @@
     "glob": "^7.1.2",
     "istanbul-api": "1.3.1",
     "istanbul-reports": "1.3.0",
-    "jest": "^23.6.0",
+    "jest": "^24.7.1",
```
Comment (dlabaj) [inline package.json]: Why did we update to 24.7.1?  Did we need something in the newer release?

File: packages/patternfly-4/react-core/src/components/Breadcrumb-2/Breadcrumb.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+---
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb-2/Breadcrumb.md]: You can just get rid of everything under Breadcrumb-2 directory.  Looks like it's the old code and the new typescript updates are in Breadcrumb

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.test.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.test.tsx]: Should be import * as React from 'react';  Believe it needs to be updated through all test.tsx files in Breadcrumb directory.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import styles from '@patternfly/patternfly/components/Breadcrumb/breadcrumb.css';
+import { css } from '@patternfly/react-styles';
+
+export interface BreadcrumbProps extends React.HTMLProps<HTMLElement> {
+  children?: React.ReactNode;
+  className?: string;
+  'aria-label'?: string;
+}
+
+const Breadcrumb: React.FunctionComponent<BreadcrumbProps> = ({
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: Should change from const Breadcrumb ... to export const Breadcrumb.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+export interface BreadcrumbHeadingProps extends React.HTMLProps<HTMLLIElement> {
+  children?: React.ReactNode;
+  className?: string;
+  to?: string;
+  target?: string;
+  component?: React.ReactType<BreadcrumbHeadingProps>;
+}
+
+const BreadcrumbHeading: React.FunctionComponent<BreadcrumbHeadingProps> = ({
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: Same comment as above export const BreadcrumbHeading.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+export interface BreadcrumbItemProps extends React.HTMLProps<HTMLLIElement> {
+  children?: React.ReactNode;
+  className?: string;
+  to?: string;
+  isActive?: boolean;
+  target?: string;
+  component?: React.ReactType<BreadcrumbItemProps>;
+}
+
+const BreadcrumbItem: React.FunctionComponent<BreadcrumbItemProps> = ({
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx]: Same as previous comment.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/index.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export { default as Breadcrumb } from './Breadcrumb';
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/index.ts]: Would change these to:

export * from './Breadcrumb';
export * from './BreadcrumbItem';
export * from './BreadcrumbHeading';

Otherwise you won't be able to access the interfaces in typescript.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import * as React from 'react';
+import styles from '@patternfly/patternfly/components/Breadcrumb/breadcrumb.css';
+import { css } from '@patternfly/react-styles';
+
+export interface BreadcrumbProps extends React.HTMLProps<HTMLElement> {
+  children?: React.ReactNode;
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: These Props are missing comments.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import styles from '@patternfly/patternfly/components/Breadcrumb/breadcrumb.css';
+import { css } from '@patternfly/react-styles';
+
+export interface BreadcrumbHeadingProps extends React.HTMLProps<HTMLLIElement> {
+  children?: React.ReactNode;
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: these props are missing comments.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import * as React from 'react';
+import { AngleRightIcon } from '@patternfly/react-icons';
+import styles from '@patternfly/patternfly/components/Breadcrumb/breadcrumb.css';
+import { css, getModifier } from '@patternfly/react-styles';
+
+export interface BreadcrumbItemProps extends React.HTMLProps<HTMLLIElement> {
+  children?: React.ReactNode;
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx]: These Props are missing comments

File: packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap [hunk 11]
```diff
@@ -13,38 +12,28 @@ exports[`Breadcrumb component should render breadcrumb with aria-label 1`] = `
...
 >
   <ol
     className="pf-c-breadcrumb__list"
   >
-    <BreadcrumbItem
-      className=""
-      component="a"
-      isActive={false}
-      target={null}
+    <Component
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap]: I don't understand why this is rendering as `<Component>` now?

File: package.json [hunk 12]
```diff
@@ -128,4 +128,4 @@
       "packages/**"
     ]
   }
-}
\ No newline at end of file
+}
```
Comment (tlabaj) [inline package.json]: can you revert this change.  it is not needed.

File: yarn.lock [hunk 13]
```diff
@@ -20886,4 +20886,4 @@ zen-observable@^0.8.0:
 
 zwitch@^1.0.0:
   version "1.0.3"
-  resolved "https://registry.yarnpkg.com/zwitch/-/zwitch-1.0.3.tgz#159fae4b3f737db1e42bf321d3423e4c96688a18"
+  resolved "https://registry.yarnpkg.com/zwitch/-/zwitch-1.0.3.tgz#159fae4b3f737db1e42bf321d3423e4c96688a18"
```
Comment (tlabaj) [inline yarn.lock]: Can you revert this change as well.  It is not needed.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap (hunk 11 above)
Comment (jessiehuff) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap]: It's rendering `component: Component = 'a' as any` in the shallow render now since it doesn't use defaultProps.

**Comment by nicolethoen** [comment]
@jessiehuff can you also add integration tests for this component as outlined on this [README](https://github.com/patternfly/patternfly-react/tree/master/packages/patternfly-4/react-integration)?

File: packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap (hunk 11 above)
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap]: If we do mount does it render 'a'?  We need to look into this further.  Rendering as `Component` here devalues the usefulness of the snapshot test in my opinion.

File: packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+describe('Breadcrumb Demo Test', () => {
```
Comment (tlabaj) [inline packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts]: You need to verify something a the component level.  With this test, we don't even  know if the component loaded.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap (hunk 11 above)
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap]: I opened an issue to track this

Comment (jessiehuff) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/__snapshots__/Breadcrumb.test.tsx.snap]: Awesome, thanks! I dug into this issue a little bit, and it seemed more complicated than I originally thought.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  'aria-label'?: string;
+}
+
+export const Breadcrumb: React.FunctionComponent<BreadcrumbProps> = ({
+  children = null,
+  className = '',
+  'aria-label': ariaLabel = '',
+  ...props
+}) => (
+  <nav {...props} className={css(styles.breadcrumb, className)}>
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: Add `aria-label={ariaLabel}` to this tag.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  /** Aria-label added to the breadcrumb nav. */
+  'aria-label'?: string;
+}
+
+export const Breadcrumb: React.FunctionComponent<BreadcrumbProps> = ({
+  children = null,
+  className = '',
+  'aria-label': ariaLabel = '',
+  ...props
+}) => (
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: Change to `}: BreadcrumbProps) => (`

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx [hunk 17]
```diff
@@ -1,33 +1,28 @@
...
+  to?: string;
   /** Target for breadcrumb link. */
-  target: PropTypes.string,
+  target?: string;
   /** Sets the base component to render. Defaults to <a> */
-  component: componentShape,
-  /** Additional props are spread to the container <li> */
-  '': PropTypes.any
-};
+  component?: React.ReactType<BreadcrumbHeadingProps>;
```
[the remark this answers is not in the document]
Comment (redallen) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: You can remove the `<BreadcrumbHeadingProps>`. This is the correct way to specify strings that are HTML elements.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx [hunk 18]
```diff
@@ -1,33 +1,28 @@
...
-
-const BreadcrumbHeading = ({ className, children, to, target, component: Component, ...props }) => (
+export const BreadcrumbHeading: React.FunctionComponent<BreadcrumbHeadingProps> = ({
+  children = null,
+  className = '',
+  to = null,
+  target = null,
+  component: Component = 'a' as any,
+  ...props
+}) => (
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: Change to `}: BreadcrumbHeadingProps) => (`

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx [hunk 19]
```diff
@@ -1,43 +1,38 @@
...
-
-const defaultProps = {
-  children: null,
-  className: '',
-  to: null,
-  isActive: false,
-  target: null,
-  component: 'a'
-};
+  component?: React.ReactType<BreadcrumbItemProps>;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx]: You can remove the `<BreadcrumbItemProps>`.

File: packages/patternfly-4/react-integration/README.md [hunk 20]
```diff
@@ -62,4 +62,4 @@ Once you’ve completed writing your integration tests, run all tests to verify
 
 Run the following command from the root directory to run all tests:
 
-`yarn run test:integration`
+`yarn  test:integration`
```
Comment (redallen) [inline packages/patternfly-4/react-integration/README.md]: You have an extra space here. Might as well fix it.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+  /** Additional classes added to the breadcrumb nav. */
+  className?: string;
+  /** Aria label added to the breadcrumb nav. */
+  'aria-label'?: string;
+}
+
+export const Breadcrumb: React.FunctionComponent<BreadcrumbProps> = ({
+  children = null,
+  className = '',
+  'aria-label': ariaLabel = '',
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: Is there a reason this does not default to Breadcrumb like the current implementations?

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx [hunk 22]
```diff
@@ -1,33 +1,28 @@
...
+  to?: string;
   /** Target for breadcrumb link. */
-  target: PropTypes.string,
+  target?: string;
   /** Sets the base component to render. Defaults to <a> */
-  component: componentShape,
-  /** Additional props are spread to the container <li> */
-  '': PropTypes.any
-};
+  component?: React.ReactType;
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: this does not match the current definition of 'componentShape`
Can you try and update the definition of component shape to
`export type componentShape = () => void | string | object;`
If that doesn't work we will investigate further.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx [hunk 23]
```diff
@@ -1,43 +1,38 @@
...
-
-const defaultProps = {
-  children: null,
-  className: '',
-  to: null,
-  isActive: false,
-  target: null,
-  component: 'a'
-};
+  component?: React.ReactType;
```
Comment (tlabaj) [inline packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx]: This should be updated based on the discussions we had yesterday.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx (hunk 22 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: This should be updated to this 
```suggestion
  component?: React.ReactType<BreadcrumbHeadingProps>;
```
@tlabaj ComponetShape is no longer needed with typescript.  It refers to Proptypes.

File: packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap [hunk 24]
```diff
@@ -149,47 +149,27 @@ exports[`Check page to verify breadcrumb is created 1`] = `
...
-      className=""
-    >
-      <BreadcrumbItem
-        className=""
-        component="a"
-        isActive={false}
-        target={null}
-        to={null}
-      >
+    <Unknown>
```
Comment (dlabaj) [inline packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap]: This seems strange.... I don't think it's suppose to be Unknown.  Can you take a look at what's going on?

File: packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+  }); 
+
+  it('Verify default section home content', () => {
+    cy.get('.pf-c-breadcrumb__link').first().contains('Section Home'); 
+  });
+
+  it('Verify default section label defaults', () => {
+    cy.get('.pf-c-breadcrumb__link').contains('Section Title'); 
+  }); 
+});
```
Comment (dlabaj) [inline packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts]: Can you add a test to click on the components in the breadcrumb and change the url?

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx (hunk 22 above)
Comment (dlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: @tlabaj ... ahh component type might be needed but then we might have to create our own type.  let me check what this should be.

Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: I am confused are you saying  that React  ReactType  covers functions, object and string types?
(Note: ReactType is deprecated we should use ElementType)

File: packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap (hunk 24 above)
Comment (jessiehuff) [reply packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap]: It seems like the Page.test.js file needed the PageBreadcrumb const to be a function. Making it `const PageBreadcrumb = () =>()` seems to fix it, not sure why.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx (hunk 22 above)
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/BreadcrumbHeading.tsx]: @jessiehuff did the above definition of componentShape not work?  What you have here will break JS consumers.

File: packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx (hunk 21 above)
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/Breadcrumb/Breadcrumb.tsx]: I still think this should default to Breadcrumb like it did before.

File: packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap [hunk 26]
```diff
@@ -512,47 +299,27 @@ exports[`Check page to verify skip to content is created 1`] = `
...
+      <Unknown>
         Section Home
-      </BreadcrumbItem>
-      <BreadcrumbItem
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Page/__snapshots__/Page.test.js.snap]: Sad to see these go, but alas our jest renderer does not support TypeScript.

File: packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts [hunk 27]
```diff
@@ -0,0 +1,24 @@
+describe('Breadcrumb Demo Test', () => {
+  it('Navigate to demo section', () => {
+    cy.visit('http://localhost:3000/');
+    cy.get('#breadcrumb-demo-nav-item-link').click();
+    cy.url().should('eq', 'http://localhost:3000/breadcrumb-demo-nav-link') 
+  });
+
+  it('Verify breadcrumb items have link', () => {
+    cy.get('li').children('a');
```
Comment (dlabaj) [inline packages/patternfly-4/react-integration/cypress/integration/breadcrumb.spec.ts]: Change to:
```suggestion
    cy.get('li').children('a').should('not.be.undefined');
    cy.get('li').children('a').should('not.equal', null);
```
